File: random/galaxy.ml
module S = Set.Make(Int)

module Arvore = struct
    type t = Arvore of node

    (* values associated with each node *)
    and node = {
        nome: int;
        valor: int;
        mutable subarv_peso: int;
        mutable cam_peso: int
    }

    (* store node nome to node *)
    let memory: (int, node) Hashtbl.t = Hashtbl.create 10000

    let tree_ds: (int, int list) Hashtbl.t = Hashtbl.create 10000

    let fazer_nodo (nome: int) (valor: int) =
        let n = {
            nome;
            valor;
            subarv_peso = 0;
            cam_peso = 0
        } in

        Hashtbl.add memory nome n;
        Arvore n

    let get tree =
        match tree with
        | Arvore n -> n

    (* this function will silently fail if the pai does not already exist in
       the arv *)
    let rec add_node (pai: int) (node: t) =
        let n = get node in
        match Hashtbl.find_opt tree_ds pai with
            | None -> Hashtbl.add tree_ds pai [n.nome]
            | Some cs -> Hashtbl.add tree_ds pai @@ n.nome :: cs

    (* this function is not purely functional since it modifies the gg
       instead of creating a new one *)
    let propagate_weights (arv: int) =
        (* acc holds weights from root to node *)
        let rec pesooo (cur: int) (acc: int) =
            let children =
                match Hashtbl.find_opt tree_ds cur with
                | Some x -> x
                | None -> []
            in
            let cur_n = Hashtbl.find memory cur in
            match children with
            | [] ->
                cur_n.subarv_peso <- 0;
                cur_n.cam_peso <- acc;
                0
            | cs ->
                let subarv_peso = List.fold_left (fun i child ->
                    let c_n = Hashtbl.find memory child in
                    c_n.cam_peso <- c_n.valor + acc;
                    let weight = c_n.valor + pesooo child (c_n.valor + acc) in
                    if weight > i then weight else i) 0 cs in
                cur_n.subarv_peso <- subarv_peso;
                subarv_peso
        in
        ignore @@ pesooo arv 0
end


(* read input from stdin *)
let read_input () =
    let num_planetas = int_of_string @@ input_line stdin in
    let num_wormholes = int_of_string @@ input_line stdin in
    let planetas = Hashtbl.create num_planetas in

    let rec read_graph n =
        let data =
            input_line stdin
            |> String.split_on_char ' '
            |> List.map int_of_string
        in
        (match data with
        | a :: b :: cost :: [] ->
            (match Hashtbl.find_opt planetas a with
            | None -> Hashtbl.add planetas a [(b, cost)]
            | Some cs -> Hashtbl.add planetas a @@ (b, cost) :: cs);
            (match Hashtbl.find_opt planetas b with
            | None -> Hashtbl.add planetas b [(a, cost)]
            | Some cs -> Hashtbl.add planetas b @@ (a, cost) :: cs)
        | _ -> failwith "data not in expected format");

        match n with
        | 1 -> planetas
        | n -> read_graph (n - 1)
    in

    (num_planetas, read_graph num_wormholes)

let () =
    let num_planetas, gg = read_input () in
    ignore @@ Arvore.fazer_nodo 1 0;

    let rec build_tree pai table seen =
        let seen = S.add pai seen in
        let children =
            Hashtbl.find table pai
            |> List.filter (fun (n, _) -> not @@ S.mem n seen)
        in
        List.iter
            (fun (node, valor) ->
                Arvore.add_node pai
                @@ Arvore.fazer_nodo node valor;
                build_tree node table seen)
            children
    in

    build_tree 1 gg S.empty;

    Arvore.propagate_weights 1;

    let rec print_ans i n =
        if i > n then ()
        else
            (let node = Hashtbl.find Arvore.memory i in
            Printf.printf "%d\n" @@ max node.subarv_peso node.cam_peso;
            print_ans (i + 1) n)
    in
    print_ans 1 num_planetas;
    flush stdout

File: random/galaxy.py
class Node:
    # name     : string
    # children : list of Node
    def __init__(self, name, weight=0):
        self.name = name
        self.weight = weight
        self.subtree_weight = 0
        self.children = []

class Tree:
    _TMP = None

    def __init__(self):
        self.root = None
        self.nodes = set()

    # node    : Node (to add)
    # to      : string (node name to connect to)
    # weight  : int (weight of edge)
    def add_node(self, node, to, weight, cur=None):
        if cur is None:
            cur = self.root

        if self.root is None:
            print('Error, please set root first!')
            sys.exit(1)

        self.nodes.add(node.name)
        self.nodes.add(to)

        if cur.name == to:
            print(f'[DEBUG] add_node: {node.name} -> {to}')
            node.weight = weight
            cur.children.append(node)
        else:
            for child in cur.children:
                ret = self.add_node(node, to, weight, child)

    def _render_helper(self, cur, f):
        for child in cur.children:
            f.write(f'    {cur.name} -> {child.name}[label="{child.weight}"]\n')
        for child in cur.children:
            self._render_helper(child, f)

    def render(self):
        with open('/tmp/galaxy.dot', 'w') as f:
            f.write('digraph {\n')
            for node in self.nodes:
                f.write(f'    {node} [label="{node} ({Tree._TMP[node].subtree_weight})"]\n')
            self._render_helper(self.root, f)
            f.write('}\n')

    def propagate_weights(self):
        def _propagate_weights(cur):
            if len(cur.children) == 0:
                # leaf node
                cur.subtree_weight = 0
            else:
                w = []
                for child in cur.children:
                    w.append(child.weight + _propagate_weights(child))
                cur.subtree_weight = max(w)

            return cur.subtree_weight

        _propagate_weights(self.root)

def read_input():
    num_planets = int(input())
    num_wormholes = int(input())
    graph = Tree()
    done = dict()
    Tree._TMP = done

    for _ in range(num_wormholes):
        A, B, cost = list(map(int, input().split(' ')))
        ANode = None
        BNode = None
        if A not in done:
            ANode = Node(A)
            done[A] = ANode
        else:
            ANode = done[A]

        if B not in done:
            BNode = Node(B)
            done[B] = BNode
        else:
            BNode = done[B]

        if ANode.name in graph.nodes:
            # A is the parent
            graph.add_node(BNode, ANode.name, cost)
        elif BNode.name in graph.nodes:
            # B is the parent
            graph.add_node(ANode, BNode.name, cost)
        else:
            # set root
            graph.root = ANode
            graph.add_node(BNode, ANode.name, cost)

    return graph

G = read_input()
G.propagate_weights()
G.render()
